programs/bpf_loader/src/syscalls.rs

```diff
@@ -2083,25 +2084,27 @@ where
                 account_indices.push(account_index);
                 accounts.push((account, None));
                 continue;
-            } else if let Some(account_ref_index) =
+            } else if let Some(caller_account_index) =
                 account_info_keys.iter().position(|key| *key == account_key)
             {
-                let account_ref = do_translate(&account_infos[account_ref_index], invoke_context)?;
+                let mut caller_account =
+                    do_translate(&account_infos[caller_account_index], invoke_context)?;
                 {
                     let mut account = account.borrow_mut();
-                    account.copy_into_owner_from_slice(account_ref.owner.as_ref());
-                    account.set_data_from_slice(account_ref.data);
-                    account.set_lamports(*account_ref.lamports);
-                    account.set_executable(account_ref.executable);
-                    account.set_rent_epoch(account_ref.rent_epoch);
+                    account.copy_into_owner_from_slice(caller_account.owner.as_ref());
+                    caller_account.original_data_len = account.data().len();
```

Comment:
This `account.data` could have already changed in an earlier CPI so we can't trust it as the original data length. The original data length has to come from the time of serializing account parameters.